### PR TITLE
[WIP] Update Travis configuration to fail on warnings/violations

### LINKED ISSUE #57: Upgrade to new travis CI
http://docs.travis-ci.com/user/migrating-from-legacy/

### PR DESCRIPTION
**Pull request by TimvdLippe**
Do not merge yet. Must make sure that this PR is compatible with #80 .

This pull request should make Travis fail on warnings generated by the plugins. Current master has all errors fixed, so we should be able to keep up with all errors quite easily now.

Checkstyle:
image
https://travis-ci.org/SERG-Delft/jpacman-framework/jobs/125239738#L1291

PMD:
image
https://travis-ci.org/SERG-Delft/jpacman-framework/jobs/125239739#L1523

Findbugs:
image
https://travis-ci.org/SERG-Delft/jpacman-framework/jobs/125239740#L1573-L1576

Also adds the container configuration of the new Travis builds.

image

Fixes #57

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -311,12 +313,6 @@
 			</plugin>
 
 			<plugin>
-				<groupId>org.codehaus.mojo</groupId>
-				<artifactId>findbugs-maven-plugin</artifactId>
-				<version>${findbugs.plugin.version}</version>
-			</plugin>
-
-			<plugin>
```
Comment (TimvdLippe) [inline pom.xml]: I removed this as it was a duplicate of https://github.com/SERG-Delft/jpacman-framework/pull/81/files#diff-600376dffeb79835ede4a0b285078036L333

**Comment by LiamClark** [comment]
I would disagree against merging this pr. Making a build fail for a check style warning devalues the worth of the build in my opinion. Build failures should flag a test failing, indication that something is functioning improperly, making build fail for trivial check style warnings acts a bit like a brittle test. 

With a brittle test at some point you can get used to having the build fail, When the build fails the idea is oh it's probably a brittle test again that doesn't matter to much. 

The same thing can happen with check style warnings. If a build failure happens for a check style error they will happen more often and can trigger the reaction of regarding a build failure as less important or simply ignored because it's "just a check style warning".

If we deeply care about keeping the repository check style warning clean, other steps can be taken.
In my opinion build failures should be about the tests.

However this is meant as a exemplar project for the students, which gives it some value here.
I wonder wat @avandeursen thinks about this.

**Comment by TimvdLippe** [comment]
Failing a build on the output of a `linter` such as eslint or jshint is common practice in frontend development. Moreover, we require students to maintain high quality code and we grade them on it. If we do not lead by example, these statements are less valued.

To make sure that the code style is uniform between different developers by using a linter such as Checkstyle, the end result is that the code is more readable. Actively notifying contributors that their code does not conform to the current standards will deter generating warnings over time. If we do not, I think the output is ignored and the plugin remains unused.

In the end, merging a pull request with failing tests or a failing build in general is bad practice.

**Comment by JWGmeligMeyling** [comment]
> We require students to maintain high quality code and we grade them on it.

Yes, but code quality is not equal to zero warnings. In fact, we challenge students to discuss the warnings they receive. And often the warning is not even valid, or not pragmatically reasonable to fix. Checkstyle apart, but PMD and Findbugs are imho mostly a learning tool: learn the flaws from your language, and be notified about smelly constructs in your system. These tools can and will however be wrong sometimes. Or you would have to boil down their configurations to the bare minimum.

**Comment by TimvdLippe** [comment]
There are two different topics of discussion in this thread:

### Plugin reports a warning 
(note: this PR does not change the usage of the plugins)
In the event of a plugin reporting a warning there are three possible actions:

1. Update the code to resolve the warning
* Remove the warning by removing the corresponding rule in the configuration file
* Suppress the warning using `@SuppressWarnings`

It is not possible to ignore the warning. Firstly because it will clutter the report over time, secondly because it requires the maintainers to create a mental map of what warnings should be ignored and lastly because new contributors will not be able to plugin at all. If the contributors sees a long list of existing warnings, he/she will not be able to see if his/her changes are introducing new warnings.

Whether the plugins are correctly configured or are too strict, that is a different topic and we should create a separate issue tracking this discussion if desired.

### Actively enforcing the plugins
This PR makes sure that the plugins are actively enforced. As a result, warnings can be observed from the moment they are introduced and resolving them will take considerably less time. Moreover, it lessens the burden of the reviewer. He/she can now rely on the automatic check, whereas previously he/she would need to generate the report themselves and communicate this back to the contributor.

<br>

I am fine with reevaluating the configuration of the tools to optimize their usage, but I do think active enforcement of these tools is required to make them useful. If you do not, the tools will only result in an increasing number of maintainability problems on the codebase.

**Comment by txsmith** [comment]
I would not be in favor of having the build fail for a linter warning. 

The first reason being that this endorses blindly following a set of rules for which we already know that they will *never* be perfect, no matter how much you slim down the strictness of the config. There will always be false positives and false negatives that cause students to get more motivated in folowing the linter than folowing common sense.

The second reason is that I think a build should not fail on things that are not mission critical. If you work for a company (or a course, for that matter) and have to stress for a deadline, having your build fail on a Checkstyle warning is at best darn annoying and at worst it encourages developers to ignore build failures (as Liam noted). It in such situations actively hinders developers from adding value to a product and denies them the posibility of making a trade-off.

The third reason I think this is a bad idea is that from an educational perspective, we want to encourage students to think of the warnings they receive, having builds fail on linter warnings does not teach them anything at best, and at worst that the linter is some holy tool never to be ignored. In my opinion having a build failure for this is only suitable in some cases when developers are experienced enough to know the reasons behind a certain configuration of the linter.

**Comment by TimvdLippe** [comment]
Given the above discussion, I will close this issue. I do want to point out that failing CI builds is best practice on the open source web development community. Given that a significant portion of students will eventually join the community, they should be aware that these practices exist.

**Pull request closed without merging**